lib/ABI/FunctionType/Conversion.cpp

```diff
@@ -243,9 +263,125 @@ TCC::tryConvertingRegisterArguments(const ArgumentRegisters &Registers) {
+                  << Delta << ".");
+      return true;
+    } else {
+      revng_log(Log,
+                "Error: The natural alignment of a type would make it "
+                "impossible to represent as CABI: there would have to be "
+                "a hole between two arguments. Abandon the conversion.");
+      // TODO: we probably want to preprocess such functions and manually
+      //       "fill" the holes in before attempting the conversion.
+      return false;
+    }
+  } else {
+    revng_log(Log,
+              "Error: The natural alignment of a type would make it "
+              "impossible to represent as CABI: the arguments (including "
+              "the padding) would have to overlap. Abandon the conversion.");
+    return false;
+  }
+}
+
+bool ArgumentDistributor::canBeNext(model::QualifiedType CurrentType,
+                                    uint64_t CurrentOffset,
+                                    uint64_t NextOffset,
+                                    uint64_t NextAlignment) {
+  revng_log(Log,
+            "Checking whether the argument #"
+              << ArgumentIndex << " can be slotted right in:\n"
+              << serializeToString(CurrentType) << "Currently "
+              << UsedGeneralPurposeRegisterCount << " general purpose and "
+              << UsedVectorRegisterCount << " vector registers are in use.");
+
+  std::optional<uint64_t> MaybeSize = CurrentType.size();
+  revng_assert(MaybeSize.has_value() && MaybeSize.value() != 0);
+
+  if (!verifyAlignment(CurrentOffset, *MaybeSize, NextOffset, NextAlignment))
+    return false;
+
+  for (const auto &Distributed : next(CurrentType)) {
+    if (Distributed.RepresentsPadding)
+      continue; // Skip padding.
+
+    if (!Distributed.Registers.empty()) {
+      revng_log(Log,
+                "Error: Because there are still available registers, "
+                "an argument cannot just be added as is - resulting CFT would "
+                "not be compatible with the original function.");
+      return false;
+    }
+
+    if (Distributed.SizeOnStack == 0) {
+      revng_log(Log,
+                "Something went very wrong: an argument uses neither registers "
+                "nor stack.");
+      return false;
+    }
+
+    // Compute the next stack offset
+    uint64_t NextStackOffset = ABI.alignedOffset(UsedStackOffset, CurrentType);
+    NextStackOffset += ABI.paddedSizeOnStack(*MaybeSize);
+    if (Distributed.SizeOnStack != NextStackOffset - UsedStackOffset) {
+      revng_log(Log,
+                "Because the stack position wouldn't match due to holes in the "
+                "stack, an argument cannot just be added as is - resulting CFT "
+                "would not be compatible.");
+      return false;
+    }
+  }
+
+  return true;
+}
+
 std::optional<llvm::SmallVector<model::Argument, 8>>
 TCC::tryConvertingStackArguments(model::QualifiedType StackArgumentTypes,
```

Comment:
These functions are really hard to understand.

If for each `if` there was some sort of example of what kind of input would take you there, it would make things so much simpler.